feat: introduce graphql linting (#2856)
Redocly/redocly-cli: docs/@v2/guides/lint-graphql.md

---
seo:
  title: Lint GraphQL with Redocly CLI
  description: Use the Redocly CLI to validate GraphQL SDL schemas, or configure rules for GraphQL.
---

# Lint GraphQL with Redocly CLI

> {% admonition type="warning" name="Experimental GraphQL support" %}

Reviewer: ```suggestion
{% admonition type="warning" name="Experimental" %}
```

This feature is at an early stage, please use with caution and send us feedback!

> GraphQL support is experimental and its behavior, rules, and configuration may change in future releases.

Reviewer: ```suggestion
This is an experimental feature. Its behavior may change in future releases.
```

{% /admonition %}

> In addition to providing lint functionality for multiple OpenAPI formats, Redocly CLI also has experimental support for GraphQL.

Reviewer: ```suggestion
In addition to providing lint functionality for multiple OpenAPI formats, Redocly CLI also supports GraphQL.
```

Redocly CLI supports the following linting approaches with GraphQL documents:

- GraphQL SDL syntax and schema validation.
- Built-in rules for checking common standards requirements (see the [list of GraphQL rules](#graphql-rules)).
- [Configurable rules](../rules/configurable-rules.md) so that you can build your own rules following common patterns.

## Lint an existing GraphQL file

Redocly CLI takes its settings from a `redocly.yaml` configuration file.

> Below is an example of a simple configuration file for validating that a GraphQL SDL file is well-formed and structurally valid:

Reviewer: ```suggestion
The following is an example of a simple configuration file that checks if a GraphQL SDL file is well-formed and has a valid structure:
```

```yaml
rules:
  struct: error
```

The `struct` rule reports an error if the SDL contains a syntax error or if the schema is structurally invalid (for example, a field that references a type that isn't defined).

With this configuration file, and your GraphQL schema file, run the linting command:

```sh
redocly lint schema.graphql
```

> The output describes any structural problems with the document, or reports that it is valid.

Reviewer: ```suggestion
The output describes structural problems with the document, or reports that it is valid.
```

{% admonition type="info" name="Syntax errors always stop linting" %}

> A GraphQL syntax error is always reported as an error and short-circuits linting for that file, regardless of how `struct` is configured.

Reviewer: ```suggestion
A GraphQL syntax error is always reported as an error and ends linting for that file, regardless of how `struct` is configured.
```

{% /admonition %}

## GraphQL rules

To expand the linting checks for a GraphQL schema, enable the built-in GraphQL rules.
Unlike the shared `struct` rule (configured under `rules`), GraphQL-specific built-in rules are configured under the `graphqlRules` section.

> The currently supported rules are:

Reviewer: ```suggestion
The supported rules are:
```

> - `no-unused-types`: Every declared type must be referenced by another type, directive, or operation (or be a root operation type: `Query`, `Mutation`, or `Subscription`). Types that are declared but never referenced are reported. If the document has no root operation type, this rule reports nothing.

Reviewer: ```suggestion
- `no-unused-types`: Every declared type must be referenced by another type, directive, or operation (or be a root operation type: `Query`, `Mutation`, or `Subscription`).
    Types that are declared but never referenced are reported.
    If the document has no root operation type, this rule reports nothing.
```

- `type-description`: Every type definition (object, interface, enum, input object, union, and scalar) must have a non-empty description.

We expect the list to expand over time, so keep checking back - and let us know if you have any requests by [opening an issue on the GitHub repo](https://github.com/Redocly/redocly-cli/issues).

To use a rule, add its name to the `graphqlRules` configuration section (or just under the `rules` section if you don't need a clear separation), and declare the severity level (either `error`, `warn`, or `off`):

```yaml
rules:
  struct: error
graphqlRules:
  type-description: warn
```

## Configurable rule example

> Redocly CLI also offers [configurable rules](../rules/configurable-rules.md) that allow you to set assertions about the document being linted.

Reviewer: ```suggestion
Redocly CLI also offers [configurable rules](../rules/configurable-rules.md) that enable you to set assertions about the document being linted.
```

---
seo:
  title: Lint GraphQL with Redocly CLI
  description: Use the Redocly CLI to validate GraphQL SDL schemas, or configure rules for GraphQL.
---

# Lint GraphQL with Redocly CLI

{% admonition type="warning" name="Experimental" %}
This is an experimental feature. Its behavior may change in future releases.
{% /admonition %}

In addition to providing lint functionality for multiple OpenAPI formats, Redocly CLI also supports GraphQL.
Redocly CLI supports the following linting approaches with GraphQL documents:

- GraphQL SDL syntax and schema validation.
- Built-in rules for checking common standards requirements (see the [list of GraphQL rules](#graphql-rules)).

> - [Configurable rules](../rules/configurable-rules.md) so that you can build your own rules following common patterns.

Reviewer: ```suggestion
- GraphQL SDL syntax and schema validation
- built-in rules for checking common standards requirements (see the [list of GraphQL rules](#graphql-rules))
- [configurable rules](../rules/configurable-rules.md) for building rules following common patterns
```

- [configurable rules](../rules/configurable-rules.md) for building rules following common patterns

## Lint an existing GraphQL file

Redocly CLI takes its settings from a `redocly.yaml` configuration file.
The following is an example of a simple configuration file that checks if a GraphQL SDL file is well-formed and has a valid structure:

```yaml
rules:
  struct: error
```

The `struct` rule reports an error if the SDL contains a syntax error or if the schema is structurally invalid (for example, a field that references a type that isn't defined).

With this configuration file, and your GraphQL schema file, run the linting command:

```sh
redocly lint schema.graphql
```

The output describes structural problems with the document, or reports that it is valid.

{% admonition type="info" name="Syntax errors always stop linting" %}
A GraphQL syntax error is always reported as an error and ends linting for that file, regardless of how `struct` is configured.
{% /admonition %}

## GraphQL rules

To expand the linting checks for a GraphQL schema, enable the built-in GraphQL rules.

> Unlike the shared `struct` rule (configured under `rules`), GraphQL-specific built-in rules are configured under the `graphqlRules` section.

Reviewer 2: Looks like we have mismatch in docs. We can also configure graphQL specific rules under `rules` section, you are listed this below.

Author: Right.

Author: Resolved.
